components/core/src/clp/ffi/ir_stream/Deserializer.cpp

```diff
@@ -0,0 +1,705 @@
+    RevertManager(SuccessHandler success_handler, FailureHandler failure_handler)
+            : m_success_handler{success_handler},
+              m_failure_handler{failure_handler} {}
+
+    // Delete copy/move constructor and assignment
+    RevertManager(RevertManager const&) = delete;
+    RevertManager(RevertManager&&) = delete;
+    auto operator=(RevertManager const&) -> RevertManager& = delete;
+    auto operator=(RevertManager&&) -> RevertManager& = delete;
+
+    // Destructor
+    ~RevertManager() {
+        if (m_success) {
+            m_success_handler();
+        } else {
+            m_failure_handler();
+        }
+    }
+
+    // Methods
+    /**
+     * Marks the state as success. By default, the object is in the failure state.
+     */
+    auto mark_as_success() -> void { m_success = true; }
+
+private:
+    // Variables
+    SuccessHandler m_success_handler;
+    FailureHandler m_failure_handler;
+    bool m_success{false};
+};
+
+/**
+ * @param ir_error_code
+ * @return Equivalent `std::errc` code indicating the same error type.
+ */
+[[nodiscard]] auto ir_error_code_to_errc(IRErrorCode ir_error_code) -> std::errc;
+
+/**
+ * @param tag
+ * @return Whether the tag represents a schema tree node.
+ */
+[[nodiscard]] auto is_schema_tree_node_tag(encoded_tag_t tag) -> bool;
+
+/**
+ * @param tag
+ * @return The corresponded schema tree node type on success.
+ * @return `std::nullopt` if the tag doesn't match to any defined schema tree node types.
+ */
+[[nodiscard]] auto schema_tree_node_tag_to_type(encoded_tag_t tag
+) -> std::optional<SchemaTreeNode::Type>;
+
+/**
+ * Deserializes the parent ID of a schema tree node.
+ * @param reader
+ * @param parent_id Outputs the deserialized result.
+ * @return `IRErrorCode::IRErrorCode_Success` on success.
+ * @return `IRErrorCode` indicating relevant errors.
+ */
+[[nodiscard]] auto deserialize_schema_tree_node_parent_id(
+        ReaderInterface& reader,
+        SchemaTreeNode::id_t& parent_id
+) -> IRErrorCode;
+
+/**
+ * Deserializes an integer value packet.
+ * @param reader
+ * @param tag
+ * @param val Outputs the deserialized value.
+ */
+[[nodiscard]] auto
+deserialize_int_val(ReaderInterface& reader, encoded_tag_t tag, value_int_t& val) -> IRErrorCode;
+
+/**
+ * Deserializes a string packet.
+ * @param reader
+ * @param tag A tag indicating the number of bytes used for the encoded length.
+ * @param deserialized_str Outputs the deserialized string.
+ * @return `IRErrorCode::IRErrorCode_Success` on success.
+ * @return `IRErrorCode` indicating relevant errors.
+ */
+[[nodiscard]] auto deserialize_string(
+        ReaderInterface& reader,
+        encoded_tag_t tag,
+        std::string& deserialized_str
+) -> IRErrorCode;
+
+/**
+ * Deserializes all the UTC offset packets until a non UTC offset packet tag is read.
+ * @param reader
+ * @param tag Inputs the current tag and outputs the last read tag.
+ * @param utc_offset Outputs the deserialized UTC offset.
```

Comment:
```suggestion
 * @param utc_offset Returns the deserialized UTC offset.
```

Our convention is `Returns`, right?